cmd/tailscale/cli/serve.go

```diff
@@ -115,31 +229,457 @@ func (e *serveEnv) runServe(ctx context.Context, args []string) error {
+func isTCPForwardingAny(sc *ipn.ServeConfig) bool {
+	if sc == nil || len(sc.TCP) == 0 {
+		return false
+	}
+	for _, h := range sc.TCP {
+		if h.TCPForward != "" {
+			return true
+		}
+	}
+	return false
 }
 
-func (e *serveEnv) runServeShowConfig(ctx context.Context, args []string) error {
+// isTCPForwardingOnPort checks serve config to see if
+// we're specifically forwarding TCP on the given port.
+func isTCPForwardingOnPort(sc *ipn.ServeConfig, port uint16) bool {
+	if sc == nil || sc.TCP == nil || sc.TCP[port] == nil {
+		return false
+	}
+	return !sc.TCP[port].HTTPS
+}
+
+// isServingWeb checks serve config to see if
+// we're serving a web handler on the given port.
+func isServingWeb(sc *ipn.ServeConfig, port uint16) bool {
+	if sc == nil || sc.Web == nil || sc.TCP == nil ||
+		sc.TCP[port] == nil || sc.TCP[port].HTTPS == false {
+		// not listening on port
+		return false
+	}
+	return true
+}
+
+func allNumeric(s string) bool {
+	for i := 0; i < len(s); i++ {
+		if s[i] < '0' || s[i] > '9' {
+			return false
+		}
+	}
+	return s != ""
+}
+
+// runServeStatus prints the current serve config.
+func (e *serveEnv) runServeStatus(ctx context.Context, args []string) error {
 	sc, err := e.getServeConfig(ctx)
 	if err != nil {
 		return err
 	}
-	j, err := json.MarshalIndent(sc, "", "  ")
+	if e.json {
+		j, err := json.MarshalIndent(sc, "", "  ")
+		if err != nil {
+			return err
+		}
+		j = append(j, '\n')
+		e.stdout().Write(j)
+		return nil
+	}
+	if sc == nil || (len(sc.TCP) == 0 && len(sc.Web) == 0 && len(sc.AllowFunnel) == 0) {
+		printf("No serve config\n")
+		return nil
+	}
+	st, err := e.getLocalClientStatus(ctx)
 	if err != nil {
 		return err
 	}
-	j = append(j, '\n')
-	e.stdout().Write(j)
+	if isTCPForwardingAny(sc) {
+		if err := printTCPStatusTree(ctx, sc, st); err != nil {
+			return err
+		}
+		printf("\n")
+	}
+	for hp := range sc.Web {
+		printWebStatusTree(sc, hp)
+		printf("\n")
+	}
+	// warn when funnel on without handlers
+	for hp, a := range sc.AllowFunnel {
+		if !a {
+			continue
+		}
+		_, portStr, _ := net.SplitHostPort(string(hp))
+		p, _ := strconv.ParseUint(portStr, 10, 16)
+		if _, ok := sc.TCP[uint16(p)]; !ok {
+			printf("WARNING: funnel=on for %s, but no serve config\n", hp)
+		}
+	}
+	return nil
+}
+
+func printTCPStatusTree(ctx context.Context, sc *ipn.ServeConfig, st *ipnstate.Status) error {
+	dnsName := strings.TrimSuffix(st.Self.DNSName, ".")
+	for p, h := range sc.TCP {
+		if h.TCPForward == "" {
+			continue
+		}
+		hp := ipn.HostPort(net.JoinHostPort(dnsName, strconv.Itoa(int(p))))
+		tlsStatus := "TLS over TCP"
+		if h.TerminateTLS != "" {
+			tlsStatus = "TLS terminated"
+		}
+		printf("┌── tcp://%s (%s, funnel on: %t)\n", hp, tlsStatus, isFunnelOn(sc, hp))
+		for _, a := range st.TailscaleIPs {
+			ipp := net.JoinHostPort(a.String(), strconv.Itoa(int(p)))
+			printf("├── tcp://%s\n", ipp)
+		}
+		printf("└───► tcp://%s\n", h.TCPForward)
```

Comment:
sure all this unicode stuff is going to work okay including in windows cmd.exe?

I think this is the first time we've done non-ASCII in the CLI

Reply:
moving to ascii for now will follow up